### PR TITLE
feat: add `GetImplicitUsersForResource` and `GetAllowedObjectConditions` APIs etc.

### PR DESCRIPTION
**Pull request by Dobmod**
- Added `GetImplicitUsersForResource`,`GetAllowedObjectConditions`,`GetAllDomains`,`GetAllRolesByDomain` to RBAC API
- Added `GetDomains`,`GetAllDomains` to RoleManager API
- Added two exception classes named `ObjConditionException` and `EmptyConditionException` 
- Copied unit tests from core library

### CODE REVIEW

File: src/Enforcer.php [hunk 1]
```diff
@@ -472,6 +474,188 @@ public function getImplicitUsersForPermission(string ...$permission): array
...
+
+                $trimPrefix = function ($string, $prefix) {
+                    if (strpos($string, $prefix) === 0) {
+                        return substr($string, strlen($prefix));
+                    } else {
+                        return $string;
+                    }
+                };
+
+                $objectConditions[] = $trimPrefix($policy[1], $prefix);
```
Comment (leeqvip) [inline src/Enforcer.php:543]: Please simplify this code
```suggestion
                $objectConditions[] = substr($policy[1], strlen($prefix);
```

File: src/Enforcer.php [hunk 2]
```diff
@@ -472,6 +474,188 @@ public function getImplicitUsersForPermission(string ...$permission): array
...
+        $permissions = [];
+        $subIndex = $this->model->getFieldIndex('p', Constants::SUBJECT_INDEX);
+        $objIndex = $this->model->getFieldIndex('p', Constants::OBJECT_INDEX);
+        $rm = $this->getRoleManager();
+
+        $isRole = [];
+        $roles = $this->getAllRoles();
+        foreach ($roles as $role) {
+            $isRole[$role] = true;
+        }
```
Comment (leeqvip) [inline src/Enforcer.php:581]: Please simplify this code:
```suggestion
        $isRole = array_flip($rules);
```

File: src/Enforcer.php [hunk 3]
```diff
@@ -472,6 +474,188 @@ public function getImplicitUsersForPermission(string ...$permission): array
...
     }
 
+    /**
+     * Convert permissions to string as a hash to deduplicate.
+     * 
+     * @param array $permissions
+     * 
+     * @return array
+     */
+    public function removeDumplicatePermissions(array $permissions): array
```
Comment (leeqvip) [inline src/Enforcer.php:484]: Please define it as a private function.

**Review by leeqvip — CHANGES_REQUESTED** [review]

File: src/Enforcer.php (hunk 2 above)
Comment (Dobmod) [reply src/Enforcer.php]: That‘s wonderful😲.

**Comment by Dobmod** [comment]
I made changes, please review.

File: src/Enforcer.php [hunk 4]
```diff
@@ -472,6 +474,174 @@ public function getImplicitUsersForPermission(string ...$permission): array
...
+    public function getImplicitUsersForResourceByDomain(string $resource, string $domain): array
+    {
+        $permissions = [];
+        $subIndex = $this->model->getFieldIndex('p', Constants::SUBJECT_INDEX);
+        $objIndex = $this->model->getFieldIndex('p', Constants::OBJECT_INDEX);
+        $domIndex = $this->model->getFieldIndex('p', Constants::DOMAIN_INDEX);
+        $rm = $this->getRoleManager();
+
+        $roles = $this->getAllRolesByDomain($domain);
+        $isRole =array_flip($roles);
```
Comment (leeqvip) [inline src/Enforcer.php:615]:
```suggestion
        $isRole = array_flip($roles);
```

**Review by leeqvip — APPROVED** [review]

**Pull request merged**